### PR TITLE
[Test] L4 complete diffusion feature test for Wan2.2 models

### PR DESCRIPTION
**Pull request by bjf-frz**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
This PR adds L4 test for Wan2.2 models, including Wan2.2-T2V-A14B-Diffusers, Wan2.2-I2V-A14B-Diffusers, Wan2.2-TI2V-5B-Diffusers

test features:
- Cache-DiT
- CFG-Parallel
- Ulysses-SP
- Tensor-Parallel
- VAE-Patch-Parallel
- HSDP
- Ring-Attn

## Test Plan
```
pytest -v -s ./tests/e2e/online_serving/test_wan22_expansion.py -m advanced_model --run-level=advanced_model
All 18 tests passed
```

## Test Result
```
================================================================================================================================================= warnings summary ==================================================================================================================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../../.venv/lib/python3.12/site-packages/torch/jit/_script.py:362: 14 warnings
  /xxx/.venv/lib/python3.12/site-packages/torch/jit/_script.py:362: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

../../.venv/lib/python3.12/site-packages/_pytest/config/__init__.py:1428
  /xxx/.venv/lib/python3.12/site-packages/_pytest/config/__init__.py:1428: PytestConfigWarning: Unknown config option: asyncio_mode
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================================================================================================== 18 passed, 17 warnings in 3242.63s (0:54:02) ====================================================================================================================================
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by yenuo26** [comment]
please attach test result in CI, before merging

**Comment by david6666666** [comment]
@yenuo26 @SamitHuang PTAL, thx

**Comment by bjf-frz** [comment]
> please attach test result in CI, before merging

done

File: tests/e2e/online_serving/test_wan22_expansion.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    cases = []
+
+    # Single-card: Cache-DiT (applies to all models)
+    for model_path, model_key in WAN22_MODELS:
+        cases.append(
+            pytest.param(
+                OmniServerParams(
+                    model=model_path,
+                    server_args=["--cache-backend", "cache_dit"],
```
Comment (fhfuih) [inline tests/e2e/online_serving/test_wan22_expansion.py]: WAN also support layerwise offloading. Can turn on this feature as well (combined in a single test case) to ensure it runs fine.

File: tests/e2e/online_serving/test_wan22_expansion.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    ("Wan-AI/Wan2.2-T2V-A14B-Diffusers", "t2v"),
+    ("Wan-AI/Wan2.2-I2V-A14B-Diffusers", "i2v"),
+    ("Wan-AI/Wan2.2-TI2V-5B-Diffusers", "ti2v"),
+]
+
+PARALLEL_CONFIGS = [
+    ("cfg_parallel", ["--cfg-parallel-size", "2"]),
+    ("ulysses_sp", ["--usp", "2"]),
+    ("tensor_parallel", ["--tensor-parallel-size", "2"]),
+    ("vae_patch", ["--vae-patch-parallel-size", "2"]),
```
Comment (fhfuih) [inline tests/e2e/online_serving/test_wan22_expansion.py]: Combine VAE patch parallel with tensor parallel to reduce test cases (see my example test case design in #1217 )

File: tests/e2e/online_serving/test_wan22_expansion.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+WAN22_MODELS = [
+    ("Wan-AI/Wan2.2-T2V-A14B-Diffusers", "t2v"),
+    ("Wan-AI/Wan2.2-I2V-A14B-Diffusers", "i2v"),
+    ("Wan-AI/Wan2.2-TI2V-5B-Diffusers", "ti2v"),
+]
+
+PARALLEL_CONFIGS = [
+    ("cfg_parallel", ["--cfg-parallel-size", "2"]),
+    ("ulysses_sp", ["--usp", "2"]),
```
Comment (fhfuih) [inline tests/e2e/online_serving/test_wan22_expansion.py:43]: Please also test ring

File: tests/conftest.py [hunk 4]
```diff
@@ -1625,49 +1712,192 @@ def send_diffusion_request(self, request_config: dict[str, Any], request_num: in
         stream = request_config.get("stream", False)
         modalities = request_config.get("modalities", omit)  # Most diffusion models don't require modalities param
         extra_body = request_config.get("extra_body", None)
+        messages = request_config.get("messages")
 
         if stream:
             raise NotImplementedError("Streaming is not currently implemented for diffusion model e2e test")
 
-        if request_num == 1:
```
Comment (fhfuih) [inline tests/conftest.py:1632]: ~~Did you accidentally remove this `if` clause to determine single request or concurrent requests? Seems now image generation requests will be routed to the `else` part, always sending concurrent requests even if there is only one request~~ Sorry, I see that you do preserve the original logic below

File: tests/conftest.py [hunk 5]
```diff
@@ -1625,49 +1712,192 @@ def send_diffusion_request(self, request_config: dict[str, Any], request_num: in
...
-            # Send single request
-            chat_completion = self.client.chat.completions.create(
-                model=request_config.get("model"),
-                messages=request_config.get("messages"),
-                extra_body=extra_body,
-                modalities=modalities,
-            )
+        if extra_body.get("num_frames", None):  # videos
```
[the remark this answers is not in the document]
Comment (fhfuih) [reply tests/conftest.py]: This AI comment may be OK to pass, because this payload is internal for tests, and `extra_body` should be always there with a video generation request.

***BUT***, since the video generation API logic is quite different, I think maybe it is clearer to define a dedicated `send_video_request`. This also prevents less elegant branching, like using `if extra_body.get("num_frames", None)` to determine if it is a video request---what if another future test needs to call this function with default num_frames value?

File: tests/conftest.py [hunk 6]
```diff
@@ -1625,49 +1712,192 @@ def send_diffusion_request(self, request_config: dict[str, Any], request_num: in
...
+            sys_prompt, user_prompt, vids, imgs, auds = extract_params_from_messages(messages)
```
Comment (fhfuih) [inline tests/conftest.py]: `request_config` (the function param) is just a custom dict structure that maximize the convenience of test execution. (It never has to resemble OpenAI API protocol anyways.) So, if your request config is too complicated such that you need to extra params from messages, then you don't need to construct an OpenAI-compliant message at the first place. 


Plus, many variables in this function return value is not used at all: sys_promtps, vids, auds.

Try simplify this request_config payload part. For example, what about a `"form_data": {...}` directly inside `request_config`? This way, you can get rid of `extract_params_from_messages`, and all the dynamic form_data construction below (those about boundary ratio and flow_shift)

File: tests/conftest.py [hunk 7]
```diff
@@ -1625,49 +1712,192 @@ def send_diffusion_request(self, request_config: dict[str, Any], request_num: in
...
-                        self.client.chat.completions.create,
-                        model=request_config.get("model"),
-                        messages=request_config.get("messages"),
-                        modalities=modalities,
-                        extra_body=extra_body,
-                    )
-                    futures.append(future)
+            try:
+                # create_and_poll includes (POST /v1/videos) and poll req (GET /v1/videos/{video_id})
+                create_url = f"{self.base_url}//v1/videos"
```
Comment (fhfuih) [inline tests/conftest.py]:
```suggestion
                create_url = f"{self.base_url}/v1/videos"
```

File: tests/conftest.py [hunk 8]
```diff
@@ -70,14 +74,91 @@ def assert_image_valid(image: Path | Image.Image, *, width: int | None = None, h
...
+        if actual_num_frames <= 0:
+            raise AssertionError(f"Invalid frame count: {actual_num_frames} (must be > 0)")
+        if actual_width <= 0 or actual_height <= 0:
+            raise AssertionError(f"Invalid dimensions: {actual_width}x{actual_height} (must be > 0)")
+        if actual_fps <= 0:
+            raise AssertionError(f"Invalid FPS: {actual_fps} (must be > 0)")
+
+        # Validate against expectations
+        if num_frames is not None:
+            expected_num_frames = ((num_frames + 3) // 4) * 4 + 1
```
[the remark this answers is not in the document]
Comment (bjf-frz) [reply tests/conftest.py]: 👍

File: tests/conftest.py [hunk 9]
```diff
@@ -1625,49 +1712,192 @@ def send_diffusion_request(self, request_config: dict[str, Any], request_num: in
...
+
+                result.success = True
+                result.videos = [video_bytes.getvalue()]
+                result.e2e_latency = time.perf_counter() - start_time
+
+                assert_diffusion_response(result, request_config, run_level=self.run_level)
+
+            except Exception as e:
+                result.success = False
+                result.error_message = f"Diffusion response processing error: {str(e)}"
```
[the remark this answers is not in the document]
Comment (bjf-frz) [reply tests/conftest.py]: 👍

File: tests/conftest.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bjf-frz) [reply tests/conftest.py]: It is currently not appropriate to determine whether the video API should be used in this scenario. As the video API differs significantly from the OpenAI API, I have refactored the logic, separating `send_video_diffusion_request` from the original `send_diffusion_request`.

File: tests/e2e/online_serving/test_wan22_expansion.py (hunk 1 above)
Comment (bjf-frz) [reply tests/e2e/online_serving/test_wan22_expansion.py]: done, already added like this:
```
server_args=["--cache-backend", "cache_dit", "--enable-layerwise-offload"],
```

File: tests/e2e/online_serving/test_wan22_expansion.py (hunk 2 above)
Comment (bjf-frz) [reply tests/e2e/online_serving/test_wan22_expansion.py]: done, added like this:
```
    ("tp_vae_patch", ["--tensor-parallel-size", "2", "--vae-patch-parallel-size", "2"]),
```

File: tests/e2e/online_serving/test_wan22_expansion.py (hunk 3 above)
Comment (bjf-frz) [reply tests/e2e/online_serving/test_wan22_expansion.py:43]: added

File: tests/conftest.py (hunk 6 above)
Comment (bjf-frz) [reply tests/conftest.py]: I have refactored the `request_confign` to include `form_data` directly, eliminating unnecessary indirection during request processing. Additionally, the asset validation logic has been streamlined: `assert_diffusion_response` now dispatches to type-specific handlers (`assert_video_diffusion_response` and `assert_image_diffusion_response`), which in turn call dedicated validators (`assert_video_valid` and `assert_image_valid`). This separation of concerns makes the structure clearer and more maintainable.

File: tests/conftest.py (hunk 7 above)
Comment (bjf-frz) [reply tests/conftest.py]: Thanks for the careful review. What's surprising, though, is that it runs successfully without throwing any errors.

File: tests/e2e/online_serving/test_wan22_expansion.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        "guidance_scale": 4.0,
+        "seed": 42,
+        # flow_shift omitted: Service uses resolution-based defaults (12.0 for 512px)
+        # vae_use_slicing/tiling omitted: Service-side optimization, not request param
+    }
+
+    if is_moe_model:
+        form_data.update(
+            {
+                "guidance_scale_high": 1.0,
```
Comment (hsliuustc0106) [inline tests/e2e/online_serving/test_wan22_expansion.py]: `/v1/videos` expects the Wan2.2 high-noise guidance field as `guidance_scale_2` (`VideoGenerationRequest` + `create_video`), but this test sends `guidance_scale_high`. That means the MoE-only branch here never actually exercises the intended high-noise CFG path even though the test still passes. Could you switch this key to `guidance_scale_2` so the I2V-A14B case validates the behavior this PR is trying to cover?

Comment (bjf-frz) [reply tests/e2e/online_serving/test_wan22_expansion.py]: fixed

File: tests/conftest.py [hunk 11]
```diff
@@ -1531,12 +1710,15 @@ def _process_diffusion_response(self, chat_completion) -> DiffusionResponse:
...
-                                    images.append(img)
+                            if isinstance(item, dict):
+                                image_url = item.get("image_url", {}).get("url")
+                            else:
+                                image_url_obj = getattr(item, "image_url", None)
+                                image_url = hasattr(image_url_obj, "url", None) if image_url_obj else None
+                            if image_url and image_url.startswith("data:image"):
+                                b64_data = image_url.split(",", 1)[1]
+                                img = decode_b64_image(b64_data)
+                                images.append(img)
```
Comment (bjf-frz) [inline tests/conftest.py:1721]: @fhfuih type of `item` here is a dict not an obj like 
```
{'type': 'image_url', 'image_url': {'url': 'data:image/png;base64,iVBxxx='}, 'stage_durations': {}}
```
I've added a dict-type check to bypass this issue or the subsequent assert would not actually be executed.

File: tests/conftest.py (hunk 7 above)
Comment (fhfuih) [reply tests/conftest.py]: Yes, it is valid URL syntax. Just not that "elegant" 😏

File: tests/conftest.py (hunk 11 above)
Comment (fhfuih) [reply tests/conftest.py:1721]: Oops thanks very much

**Review by fhfuih — APPROVED** [review]
LGTM

**Comment by david6666666** [comment]
@yenuo26  PTAL thx

**Comment by david6666666** [comment]
LGTM now

**Comment by fhfuih** [comment]
Per our offline discussion with @Gaohan123 & @congw729 , could you try somehow redesign the feature & modality combinations and reduce the number of test cases to save some time? Wan 2.2 indeed has covered many modalities, making the test design more complex

**Comment by bjf-frz** [comment]
> Per our offline discussion with @Gaohan123 & @congw729 , could you try somehow redesign the feature & modality combinations and reduce the number of test cases to save some time? Wan 2.2 indeed has covered many modalities, making the test design more complex

Case assignment has been configured to be randomized, with a random seed generated each time based on the current system timestamp.

**Comment by david6666666** [comment]
LGTM and test passed.

File: .buildkite/test-nightly.yml [hunk 12]
```diff
@@ -111,6 +111,47 @@ steps:
                   path: /mnt/hf-cache
                   type: DirectoryOrCreate
 
+  - label: ":full_moon: Diffusion Model Wan22 completed Test with H100"
+    timeout_in_minutes: 45
+    depends_on: upload-nightly-pipeline
+    if: build.env("NIGHTLY") == "1" || build.pull_request.labels includes "nightly-test"
+    commands:
+      - export VLLM_WORKER_MULTIPROC_METHOD=spawn
+      - pytest -s -v tests/e2e/online_serving/test_wan22_expansion.py -m "advanced_model" --run-level "advanced_model"
```
Comment (yenuo26) [inline .buildkite/test-nightly.yml:120]: I think just modifying here maybe cause both "Diffusion Model Wan22 completed Test with H100" and "Diffusion Model Test with H100" to run this test case.

Comment (bjf-frz) [reply .buildkite/test-nightly.yml:120]: I removed the diffusion mark in the wan22 test script, it won't run in `Diffusion Model Test with H100`.

Comment (yenuo26) [reply .buildkite/test-nightly.yml:120]: Maybe you can change to use pytest --ignore in subsequent PR? I think not adding the diffusion tag may cause some statistical issues later.

Comment (bjf-frz) [reply .buildkite/test-nightly.yml:120]: ok
